diff --git a/src/gl_calls.c b/src/gl_calls.c
--- a/src/gl_calls.c
+++ b/src/gl_calls.c
@@ -2504,7 +2504,11 @@ void glGetInteger64i_v(GLenum target, GLuint index, GLint64* data) {
 }

 void glGetInteger64v(GLenum pname, GLint64* data) {
-    println(MSG_DEBUG_UNIMPLEMENTED_GLCALL, "glGetInteger64v");
+    GL_CALL_LOCK();
+    GL_SEND_CHECKED(REQUEST_CODE_GL_GET_INTEGER64V, &pname, sizeof(GLenum));
+    GL_RECV_CHECKED();
+    *data = ArrayBuffer_getLong(&inputBuffer);
+    GL_CALL_UNLOCK();
 }

 void glGetInteger64vEXT(GLenum pname, GLint64* data) {
@@ -2514,7 +2518,13 @@ void glGetInteger64vEXT(GLenum pname, GLint64* data) {
 }

 void glGetIntegeri_v(GLenum target, GLuint index, GLint* data) {
-    println(MSG_DEBUG_UNIMPLEMENTED_GLCALL, "glGetIntegeri_v");
+    uint64_t request = (uint64_t)target | ((uint64_t)index << 32);
+    GL_CALL_LOCK();
+    GL_SEND_CHECKED(REQUEST_CODE_GL_GET_INTEGERI_V, &request,
+                    sizeof(request));
+    GL_RECV_CHECKED();
+    *data = ArrayBuffer_getInt(&inputBuffer);
+    GL_CALL_UNLOCK();
 }

 void glGetIntegeri_vEXT(GLenum target, GLuint index, GLint* data) {
@@ -2916,7 +2920,17 @@ void glGetShaderInfoLog(GLuint shader, GLsizei bufSize, GLsizei* length, GLchar*
 }

 void glGetShaderPrecisionFormat(GLenum shadertype, GLenum precisiontype, GLint* range, GLint* precision) {
-    println(MSG_DEBUG_UNIMPLEMENTED_GLCALL, "glGetShaderPrecisionFormat");
+    GL_CALL_LOCK();
+    ArrayBuffer_rewind(&outputBuffer);
+    ArrayBuffer_putInt(&outputBuffer, shadertype);
+    ArrayBuffer_putInt(&outputBuffer, precisiontype);
+    GL_SEND_CHECKED(REQUEST_CODE_GL_GET_SHADER_PRECISION_FORMAT,
+                    outputBuffer.buffer, outputBuffer.size);
+    GL_RECV_CHECKED();
+    range[0] = ArrayBuffer_getInt(&inputBuffer);
+    range[1] = ArrayBuffer_getInt(&inputBuffer);
+    *precision = ArrayBuffer_getInt(&inputBuffer);
+    GL_CALL_UNLOCK();
 }

 void glGetShaderSource(GLuint shader, GLsizei bufSize, GLsizei* length, GLchar* source) {
