diff --git a/include/gl_context.h b/include/gl_context.h
--- a/include/gl_context.h
+++ b/include/gl_context.h
@@ -39,7 +39,7 @@ typedef struct __GLXFBConfigRec {
 } __GLXFBConfigRec;

 extern bool setupRingBuffers(GLContext* context);
-extern bool gladioInitOnce();
+extern bool gladioInitOnce(Display* dpy);
 extern char* getCachedString(int name);
 extern char* putCachedString(int name, char* string, int length);
 extern void writeUnboundVertexArrays(GLint first, GLsizei count, const void* indices, GLenum indexType, GLint basevertex);
diff --git a/include/gladio.h b/include/gladio.h
--- a/include/gladio.h
+++ b/include/gladio.h
@@ -7,11 +7,11 @@
 #define DEBUG_MODE (0) // FIXME 0=DISABLE or (FIRST | SECOND)
 #define IS_DEBUG_ENABLED(type) ((DEBUG_MODE & type) != 0)

-#define GLX_EXT_MAJOR_OPCODE -106
 #define SERVER_RING_BUFFER_SIZE 67108864
 #define CLIENT_RING_BUFFER_SIZE 33554432
 #define THREAD_POOL_NUM_THREADS 4
 #define PIXEL_READ_CACHE_SKIP_FRAMES 3
+extern char glxMajorOpcode;
 #define SKIP_GL_FINISH 1
 #define X11_SERVER_PATH "/data/data/com.winlator/files/rootfs/tmp/.X11-unix/X0"

@@ -148,7 +148,7 @@ static inline bool gl_recv(RingBuffer* ring, short* outRequestCode, ArrayBuffer*

 static inline bool glx_send(int fd, char opcode, void* data, int size) {
     char header[4];
-    header[0] = GLX_EXT_MAJOR_OPCODE;
+    header[0] = glxMajorOpcode;
     header[1] = opcode;
     *(short*)(header + 2) = size / 4 + 1;

diff --git a/src/glx_calls.c b/src/glx_calls.c
--- a/src/glx_calls.c
+++ b/src/glx_calls.c
@@ -30,7 +30,7 @@ void glXCopyContext(Display* dpy, GLXContext src, GLXContext dst, unsigned long
 }

 GLXContext glXCreateContextAttribsARB(Display* dpy, GLXFBConfig config, GLXContext share_context, Bool direct, const int* attrib_list) {
-    if (!gladioInitOnce()) return NULL;
+    if (!gladioInitOnce(dpy)) return NULL;
     GLX_CALL_LOCK();
     int contextId = maxContextId++;

@@ -74,7 +74,7 @@ GLXContext glXCreateContextAttribsARB(Display* dpy, GLXFBConfig config, GLXConte
 }

 GLXContext glXCreateContext(Display* dpy, XVisualInfo* vis, GLXContext shareList, Bool direct) {
-    if (!gladioInitOnce()) return NULL;
+    if (!gladioInitOnce(dpy)) return NULL;
     GLX_CALL_LOCK();
     int contextId = maxContextId++;

diff --git a/src/main.c b/src/main.c
--- a/src/main.c
+++ b/src/main.c
@@ -9,6 +9,7 @@
 int serverFd = -1;
 ArrayBuffer outputBuffer = {0};
 ArrayBuffer inputBuffer = {0};
+char glxMajorOpcode = 0;
 SparseArray glxContexts = {0};
 thread_local GLContext* currentGLContext = NULL;

@@ -197,7 +198,15 @@ static bool sendX11AuthRequest() {
     return true;
 }

-bool gladioInitOnce() {
+bool gladioInitOnce(Display* dpy) {
+    if (glxMajorOpcode == 0) {
+        int majorOpcode;
+        int firstEvent;
+        int firstError;
+        if (!XQueryExtension(dpy, "GLX", &majorOpcode, &firstEvent,
+                             &firstError)) return false;
+        glxMajorOpcode = (char)majorOpcode;
+    }
     if (serverFd == -1) {
         serverFd = gladioServerConnect();
